util/network/src/client_version.rs

```diff
@@ -0,0 +1,473 @@
+/// "Geth/main.jnode.network/v1.8.21-stable-9dc5d1a9/linux-amd64/go1.11.4"
+
+#[derive(Clone,Debug,PartialEq,Eq)]
+pub struct ParityClientData {
+	name: String,
+	variant: Option<String>,
+	semver: Version,
+	os: String,
+	compiler: String,
+}
+
+// Accessor methods for ParityClientData. This will probably
+// need to be abstracted away into a trait.
+impl ParityClientData {
+	fn name(&self) -> &str {
+		self.name.as_str()
+	}
+
+	fn variant(&self) -> Option<&str> {
+		self.variant.as_ref().map(String::as_str)
+	}
+
+	fn semver(&self) -> &Version {
+		&self.semver
+	}
+
+	fn os(&self) -> &str {
+		self.os.as_str()
+	}
+
+	fn compiler(&self) -> &str {
+		self.compiler.as_str()
+	}
+}
+
+#[derive(Clone,Debug,PartialEq,Eq)]
+pub enum ClientVersion {
+	ParityClient(
+		ParityClientData
+	),
+	ParityUnknownFormat(String),
+	Other(String), // Id string
+}
+
+// TODO: Maybe merge with Peercapabilityinfo in ethcore-network?
+pub trait ClientCapabilities {
+	fn can_handle_large_requests(&self) -> bool;
+
+	fn accepts_service_transaction(&self) -> bool;
+}
+
+// This is an implementation of a function taken from propagator.rs
+fn parity_accepts_service_transaction(parity_client_data: &ParityClientData) -> bool {
+	let service_transactions_version = Version::parse(SERVICE_TRANSACTIONS_VERSION).unwrap();
+
+	*parity_client_data.semver() >= service_transactions_version
+}
+
+impl ClientCapabilities for ClientVersion {
+	fn can_handle_large_requests(&self) -> bool {
+		match self {
+			ClientVersion::ParityClient(data) => {
+				if *data.semver() < Version::parse(PARITY_CLIENT_LARGE_REQUESTS_VERSION).unwrap() {
+					false
+				} else {
+					true
+				}
+			},
+			ClientVersion::ParityUnknownFormat(_) => false, // Play it safe
+			ClientVersion::Other(_) => true // As far as we know
+		}
+	}
+
+	/// Checks if peer is able to process service transactions
+	fn accepts_service_transaction(&self) -> bool {
+		match self {
+			ClientVersion::ParityClient(data) => parity_accepts_service_transaction(&data),
+			ClientVersion::ParityUnknownFormat(_) => false,
+			ClientVersion::Other(_) => false
+		}
+	}
+
+}
+
+
+fn is_parity(client_id: &str) -> bool {
+	if client_id.starts_with(LEGACY_CLIENT_ID_PREFIX) || client_id.starts_with(PARITY_CLIENT_ID_PREFIX) {
+		return true;
+	} else {
+		return false;
+	}
+}
+
+// Parse known parity formats.
+//
+// This is really not robust: parse four arguments and
+// allow for an extra argument between identifier and
+// version
+// TODO implement a better logic
+fn parse_parity_format(client_version: &str) -> Result<ParityClientData, ()> {
+	let tokens: Vec<&str> = client_version.split("/").collect();
+
+
+	// Basically strip leading 'v'
+	if let Some(version_number) = &get_number_from_version(tokens[1]) {
+		if tokens.len() < 4 {
+			return Err(());
+		}
+
+		return Ok(
+			ParityClientData {
+				name: tokens[0].to_string(),
+				variant: None,
+				semver: Version::parse(version_number).unwrap(),
```

Comment:
Please either use `expect()` or turn into error (I think the latter should be done).